src/backend/distributed/planner/distributed_planner.c

```diff
@@ -703,11 +703,20 @@ CreateDistributedPlan(uint64 planId, Query *originalQuery, Query *query, ParamLi
 		Query *newQuery = copyObject(originalQuery);
 		bool setPartitionedTablesInherited = false;
 
-		/* remove the pre-transformation planner restrictions context */
-		PopPlannerRestrictionContext();
+		PG_TRY();
+		{
+			/* remove the pre-transformation planner restrictions context */
+			PopPlannerRestrictionContext();
+
+			/* create a fresh new planner context */
+			plannerRestrictionContext = CreateAndPushPlannerRestrictionContext();
```

Comment:
In principle this works, but I think it would be nicer to define a function to reset the current planner restriction context, e.g.:
```c
static void
ResetPlannerRestrictionContext(PlannerRestrictionContext *plannerRestrictionContext)
{
    plannerRestrictionContext->relationRestrictionContext =
        palloc0(sizeof(RelationRestrictionContext));

    plannerRestrictionContext->joinRestrictionContext =
        palloc0(sizeof(JoinRestrictionContext));

    plannerRestrictionContext->relationRestrictionContext->allReferenceTables = true;
}
```